Cardinality check for metrics

rexagod opened the pull request:
Allow reduced cardinality for metrics, so that the overall influx of
data reaching data-hub's telemeter is minimal.
* introduces new wrapper metrics to avoid collision
* disallows parsing original noobaa rules and uses custom rules instead,
  to avoid conflict with products dependent on noobaa metrics, ODF, for
  instance.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented:
/hold
Need to run this by @cloudbehl first.

jmolmo commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,211 @@
+           severity: critical
+       - alert: NooBaaSystemCapacityWarning100
+         annotations:
+           description: A NooBaa system approached its capacity, usage is at 100%
```
Maybe we do not need this alert..Are we sure that the capacity can be used completely, I mean, probably we are going to have always some bytes "free". In any case it does not hurt ... this is a simple comment.

jmolmo commented on controllers/monitoring_alert.go:
```diff
@@ -180,7 +213,15 @@ func (r *ManagedMCGReconciler) reconcileMonitoringResources() error {
+			r.Log.Info("found noobaa prometheus rules, checking for inclusion")
+			if v, ok := obj.GetLabels()[monLabelKey]; ok && v == monLabelValue {
+				utils.RemoveLabel(obj, monLabelKey)
+				r.Log.Info("removing app label from noobaa prometheus rules")
```
IMHO, a single quote will make more easy to understand the message. ---> "removing 'app' label", instead "removing app label"

jmolmo commented on controllers/monitoring_alert.go:
```diff
@@ -180,7 +213,15 @@ func (r *ManagedMCGReconciler) reconcileMonitoringResources() error {
 	for i := range promRuleList.Items {
 		obj := promRuleList.Items[i]
-		utils.AddLabel(obj, monLabelKey, monLabelValue)
+		if obj.GetName() == "noobaa-prometheus-rules" {
```
It seems that this is just a "cleaning" task to revert a label set in previous versions. If i am not wrong, maybe a brief comment explaining "why" we do this will make the code more easy to understand.

rexagod commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,191 @@
+       - expr: |
+           NooBaa_num_namespace_resources{namespace="redhat-data-federation"}
+         record: nooBaa_num_namespace_resources
+#       - expr: |
+#           NooBaa_namespace_resource_status{namespace="redhat-data-federation"}
+#         record: noobaa_namespace_resource_status
+#       - expr: |
+#           NooBaa_namespace_bucket_status{namespace="redhat-data-federation"}
+#         record: noobaa_namespace_bucket_status
```
Made changes based on @jmolmo's review and a conversation with @cloudbehl on Friday. I've commented this out right now as I need to test these two metrics on a cluster before I incorporate them.

rexagod replied:
@cloudbehl For these two metrics, I'm not getting anything in Prometheus.

<details>
<summary>Screenshots</summary>

![Screenshot from 2022-07-13 13-29-28](https://user-images.githubusercontent.com/33557095/178683616-38147828-9c4f-47da-ad52-791f1a647e6b.png)

![image](https://user-images.githubusercontent.com/33557095/178682278-25a139f8-4003-491c-b44e-fec18cf585de.png)

</details>

cloudbehl replied:
did you create the namespace bucket  and bucket class to see the availability of the metrics?

rexagod replied:
Yep, I created these and now they are available (as seen in the PDF I sent you). I've added them.

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,191 @@
+           system_type: NooBaa
+           system_vendor: Red Hat
+         record: noobaa_system_iops_total_bytes
+       - expr: |
+           sum by (namespace, managedBy, job, service) (rate(NooBaa_providers_bandwidth_read_size{namespace="redhat-data-federation"}[5m]) + rate(NooBaa_providers_bandwidth_write_size{namespace="redhat-data-federation"}[5m]))
+         labels:
+           system_type: NooBaa
+           system_vendor: Red Hat
+         record: noobaa_system_throughput_total_bytes
```
Is it related to namespace buckets?

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,191 @@
+         record: noobaa_system_throughput_total_bytes
+   - name: bucket-state-alert.rules
+     rules:
+       - alert: NooBaaBucketErrorState
+         annotations:
+           description: A NooBaa bucket {{ $labels.bucket_name }} is in error state for
+             more than 5m
+           message: A NooBaa Bucket Is In Error State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_bucket_status{bucket_name=~".*"} == 0
+         for: 5m
+         labels:
+           severity: warning
```
Why do we need this in DF?

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,191 @@
+         for: 5m
+         labels:
+           severity: warning
+       - alert: NooBaaBucketExceedingQuotaState
+         annotations:
+           description: A NooBaa bucket {{ $labels.bucket_name }} is exceeding its quota
+             - {{ printf "%0.0f" $value }}% used
+           message: A NooBaa Bucket Is In Exceeding Quota State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_bucket_quota{bucket_name=~".*"} >= 100
+         for: 5m
+         labels:
+           severity: warning
```
same

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,191 @@
+         for: 5m
+         labels:
+           severity: warning
+       - alert: NooBaaBucketLowCapacityState
+         annotations:
+           description: A NooBaa bucket {{ $labels.bucket_name }} is using {{ printf
+             "%0.0f" $value }}% of its capacity
+           message: A NooBaa Bucket Is In Low Capacity State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_bucket_capacity{bucket_name=~".*"} > 80
+         for: 5m
+         labels:
+           severity: warning
+       - alert: NooBaaBucketNoCapacityState
+         annotations:
+           description: A NooBaa bucket {{ $labels.bucket_name }} is using all of its
+             capacity
+           message: A NooBaa Bucket Is In No Capacity State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_bucket_capacity{bucket_name=~".*"} > 95
+         for: 5m
+         labels:
+           severity: warning
+   - name: resource-state-alert.rules
```
which buckets are these, is it backing store or NooBaa internal buckets?

rexagod commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,191 @@
+ groups:
+   - name: noobaa-telemeter.rules
```
That's outdated. Please refer [this](https://docs.google.com/document/d/19JST--IQyxTa_330D7mjJZUTEi2dqJ9NXcJ9J0SrhyY/edit#).

rexagod commented:
@cloudbehl

- Removed all `.*Quota.*` alerts.
- Removed all `.*Capacity.*` alerts.
- Removed IOPS and throughput metrics.
- Added metrics for buckets and nss.

Please review.

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,81 @@
+   - name: bucket-state-alert.rules
+     rules:
+       - alert: NooBaaNamespaceBucketErrorState
+         annotations:
+           description: A NooBaa namespace bucket {{ $labels.bucket_name }} is in error
+             state for more than 5m
+           message: A NooBaa Namespace Bucket Is In Error State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_namespace_bucket_status{bucket_name=~".*"} == 0
+         for: 5m
+         labels:
+           severity: warning
+   - name: resource-state-alert.rules
+     rules:
+       - alert: NooBaaResourceErrorState
+         annotations:
+           description: A NooBaa resource {{ $labels.resource_name }} is in error state
+             for more than 5m
+           message: A NooBaa Resource Is In Error State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_resource_status{resource_name=~".*"} == 0
+         for: 5m
+         labels:
+           severity: warning
+       - alert: NooBaaNamespaceResourceErrorState
+         annotations:
+           description: A NooBaa namespace resource {{ $labels.namespace_resource_name
+             }} is in error state for more than 5m
+           message: A NooBaa Namespace Resource Is In Error State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_namespace_resource_status{namespace_resource_name=~".*"} == 0
+         for: 5m
+         labels:
+           severity: warning
```
Instead of calling it namespace resource and bucket can we change the naming as per the product?

Bucket Class = Buckey Policy

Namespace Store = Data resource

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,81 @@
+         record: noobaa_system_health_status
+   - name: bucket-state-alert.rules
+     rules:
+       - alert: NooBaaNamespaceBucketErrorState
```
Can we name it like this?
```suggestion
       - alert:BucketPolicyErrorState
```

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,81 @@
+     rules:
+       - alert: NooBaaNamespaceBucketErrorState
+         annotations:
+           description: A NooBaa namespace bucket {{ $labels.bucket_name }} is in error
+             state for more than 5m
```
```suggestion
           description: A data source {{ $labels.bucket_name }} is in error
             state for more than 5m
```

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,81 @@
+         annotations:
+           description: A NooBaa namespace bucket {{ $labels.bucket_name }} is in error
+             state for more than 5m
+           message: A NooBaa Namespace Bucket Is In Error State
```
```suggestion
           message: A Data Source Is In Error State
```

cloudbehl reviewed (commented):
All metrics looks good.



Can we make changes to the alert as per the suggestion I provided for the first alert?

rexagod commented:
Context for the change introduced in the last commit: https://github.com/red-hat-storage/mcg-osd-deployer/pull/61/commits/7aa6b963eb9ec6ad13be73d0f2d3cc1290d7e4f5

![Screenshot from 2022-07-17 20-22-01](https://user-images.githubusercontent.com/33557095/179404092-184c900d-f5e9-4f08-b9ea-d2dd8b449f09.png)

rexagod commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,105 @@
+           origin: mcg-osd-deployer
+   - name: resource-state-alert.rules
+     rules:
+       - alert: DefaultBackingStoreErrorState
```
```suggestion
       - alert: CacheBucketErrorState
```

rexagod commented:
/cc @cloudbehl @SanjalKatiyar @jmolmo 
PTAL and approve if LGTY, thanks.

SanjalKatiyar commented:
LGTM

cloudbehl commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,105 @@
+         labels:
+           severity: warning
+           origin: mcg-osd-deployer
+       - alert: NamespaceStoreErrorState
+         annotations:
+           description: A Namespace Store {{ $labels.namespace_resource_name
+             }} is in error state for more than 5m
+           message: A Namespace Store Is In Error State
+           severity_level: warning
+           storage_type: NooBaa
+         expr: |
+           NooBaa_namespace_resource_status{namespace_resource_name=~".*"} == 0
+         for: 5m
+         labels:
+           severity: warning
+           origin: mcg-osd-deployer
```
I think we should fix the Namespace store usage to Data resource as well?

rexagod commented:
/cc @naveenpaul1
PTAL and LGTM if this looks fine to you, thanks.

naveenpaul1 commented on controllers/monitoring_alert.go:
```diff
@@ -180,7 +213,17 @@ func (r *ManagedMCGReconciler) reconcileMonitoringResources() error {
+		if obj.GetName() == "noobaa-prometheus-rules" {
+			r.Log.Info("found noobaa prometheus rules, checking for inclusion")
+			if v, ok := obj.GetLabels()[monLabelKey]; ok && v == monLabelValue {
+				utils.RemoveLabel(obj, monLabelKey)
```
one doubt, we are removing the original Prometheus rule to add the labels origin: mcg-osd-deployer which will help us to identify our metrics? if yes where are we adding this labels to metrics? @rexagod

rexagod replied:
We are removing the original rule file to remove/add/edit a bunch of custom rules and alerts, in order to provide clarity to the user. We are adding these labels here: https://github.com/red-hat-storage/mcg-osd-deployer/pull/61/files#diff-c2d8ea863a8b41a65ff336525a37d8a9b198d683b1dae92c7d4be6f06c446d7eR7.

rexagod commented:
/retest ci/prow/lint

rexagod commented:
/test lint

rexagod commented:
/test lint

rexagod commented:
/override ci/prow/lint

simonpasquier commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,105 @@
+         annotations:
+           description: A Bucket Policy {{ $labels.bucket_name }} is in error
+             state for more than 5m
+           message: A Bucket Policy Is In Error State
```
This should be `summary`.

simonpasquier commented on controllers/noobaa-rules.yaml:
```diff
@@ -0,0 +1,105 @@
+             state for more than 5m
+           message: A Bucket Policy Is In Error State
+           severity_level: warning
+           storage_type: NooBaa
```
this looks more like a label though this could also be conveyed in the alert name (e.g. `BucketPolicyErrorState` -> `NooBaaBucketPolicyErrorState`)
